src/main/java/com/microfocus/application/automation/tools/octane/actions/UFTTestDetectionPublisher.java

```diff
@@ -39,109 +40,108 @@
+		return workspaceName;
+	}
+
+	public String getScmRepositoryId() {
+		return scmRepositoryId;
+	}
+
+	// Fields in config.jelly must match the parameter names in the "DataBoundConstructor"
+	@DataBoundConstructor
+	public UFTTestDetectionPublisher(String sharedSpaceId, String workspaceName, String scmRepositoryId) {
+		this.sharedSpaceId = sharedSpaceId;
+		this.workspaceName = workspaceName;
+		this.scmRepositoryId = scmRepositoryId;
+	}
+
+	@Override
+	public boolean perform(AbstractBuild build, Launcher launcher, BuildListener listener) {
+		UftTestDiscoveryResult results = UFTTestDetectionService.startScanning(build, getWorkspaceName(), getScmRepositoryId(), listener);
+		UFTTestDetectionBuildAction buildAction = new UFTTestDetectionBuildAction(build, results);
+		build.addAction(buildAction);
+
+		return true;
+	}
+
+	@Override
+	public DescriptorImpl getDescriptor() {
+		return (DescriptorImpl) super.getDescriptor();
+	}
+
+	@Override
+	public BuildStepMonitor getRequiredMonitorService() {
+		return BuildStepMonitor.NONE;
+	}
+
+	@Extension // This indicates to Jenkins that this is an implementation of an extension point.
+	public static final class DescriptorImpl extends BuildStepDescriptor<Publisher> {
+
+		public ListBoxModel doFillSharedSpaceIdItems() {
+			ListBoxModel m = new ListBoxModel();
+			if (!ConfigurationService.getAllSettings().isEmpty()) {
+				ConfigurationService.getAllSettings().forEach(settings -> m.add(settings.getSharedSpace(), settings.getSharedSpace()));
+			}
+			return m;
+		}
+
+		public FormValidation doCheckSharedSpaceId(@QueryParameter String value) {
+			if (value == null || value.isEmpty()) {
+				return FormValidation.error("Please select shared space");
+			} else {
+				return FormValidation.ok();
+			}
+		}
+
+		public ListBoxModel doFillWorkspaceNameItems(@QueryParameter String sharedSpaceId) {
+			ListBoxModel m = new ListBoxModel();
+			if (sharedSpaceId != null && !sharedSpaceId.isEmpty()) {
+				EntitiesService entitiesService = OctaneSDK.getClientBySharedSpaceId(sharedSpaceId).getEntitiesService();
+				List<Entity> workspaces = entitiesService.getEntities(null, "workspaces", null, null);
+				for (Entity workspace : workspaces) {
+					m.add(workspace.getName(), String.valueOf(workspace.getId()));
+				}
+			}
+			return m;
+		}
+
+		public FormValidation doCheckWorkspaceName(
+				@QueryParameter(value = "workspaceName") String value,
+				@QueryParameter(value = "sharedSpaceId") String sharedSpaceId) {
+			if (sharedSpaceId == null || sharedSpaceId.isEmpty()) {
+				return FormValidation.error("Please select shared space");
+			}
+			if (value == null || value.isEmpty()) {
+				return FormValidation.error("Please select workspace");
+			}
+			return FormValidation.ok();
+		}
+
+		public boolean isApplicable(Class<? extends AbstractProject> aClass) {
+			return aClass.equals(FreeStyleProject.class);
+		}
+
+		public String getDisplayName() {
+			return "ALM Octane UFT Tests Scanner";
```

Comment:
Wouldn't it be better to move important constants to Messages.properties?

Reply:
we will fix it in next PR